commit fb9eb027fbc90e21cbda18379d85f8603e1f28b9
Author: Horatiu Vultur <horatiu.vultur@microchip.com>
Date:   Thu Mar 10 09:40:05 2022 +0100

    net: lan966x: Improve the CPU TX bitrate.
    
    When doing manual injection of the frame, it is required to check if the
    TX FIFO is ready to accept the next word of the frame. For this we are
    using 'readx_poll_timeout_atomic', the only problem is that before it
    actually checks the status, is determining the time when to finish polling
    the status. Which seems to be an expensive operation.
    Therefore check the status of the TX FIFO before calling
    'readx_poll_timeout_atomic'.
    Doing this will improve the TX bitrate by ~70%. Because 99% the FIFO is
    ready by that time. The measurements were done using iperf3.
    
    Before:
    [ ID] Interval           Transfer     Bitrate         Retr
    [  5]   0.00-10.03  sec  55.2 MBytes  46.2 Mbits/sec    0 sender
    [  5]   0.00-10.04  sec  53.8 MBytes  45.0 Mbits/sec      receiver
    
    After:
    [ ID] Interval           Transfer     Bitrate         Retr
    [  5]   0.00-10.10  sec  95.0 MBytes  78.9 Mbits/sec    0 sender
    [  5]   0.00-10.11  sec  95.0 MBytes  78.8 Mbits/sec      receiver
    
    Signed-off-by: Horatiu Vultur <horatiu.vultur@microchip.com>
    Reviewed-by: Andrew Lunn <andrew@lunn.ch>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/drivers/net/ethernet/microchip/lan966x/lan966x_main.c b/drivers/net/ethernet/microchip/lan966x/lan966x_main.c
index 81c01665d01e..e1bcb28039dc 100644
--- a/drivers/net/ethernet/microchip/lan966x/lan966x_main.c
+++ b/drivers/net/ethernet/microchip/lan966x/lan966x_main.c
@@ -185,6 +185,9 @@ static int lan966x_port_inj_ready(struct lan966x *lan966x, u8 grp)
 {
 	u32 val;
 
+	if (lan_rd(lan966x, QS_INJ_STATUS) & QS_INJ_STATUS_FIFO_RDY_SET(BIT(grp)))
+		return 0;
+
 	return readx_poll_timeout_atomic(lan966x_port_inj_status, lan966x, val,
 					 QS_INJ_STATUS_FIFO_RDY_GET(val) & BIT(grp),
 					 READL_SLEEP_US, READL_TIMEOUT_US);